README.md

```diff
@@ -29,28 +42,133 @@ running `sbcl` and evaluating `ql:*local-project-directories*`. Once
+*MAGICL extensions*. These extensions typically install new backends
+to MAGICL functions. The available extensions are:
+
+- `MAGICL/EXT-BLAS`: for BLAS functions
+- `MAGICL/EXT-LAPACK`: for LAPACK functions
+- `MAGICL/EXT-EXPOKIT`: for expokit (matrix `exp()`) functions
+
+For backwards compatibility, `MAGICL` loads every extension under the
+kitchen sink. **This may change in future versions of MAGICL!**
+
+If you use extensions, you'll need the requisite C/Fortran
+libraries. Expokit will automatically build for you, as its source is
+included in the distribution of MAGICL.
+
+### Backends
+
+Accelerated functionality is installed with a notion called "backends". A
+*backend* is a name of a group of functionality, typically denoted by
+a symbol or keyword. The `:lisp` backend is the defualt one, and
+several backends can be active all at once. Each extension above adds
+a new backend. The current backends are:
+
+- `:lisp`: Pure Lisp code
+- `:blas`: BLAS-backed code
+- `:lapack`: LAPACK-backed code
+- `:expokit`: expokit-backed code
+
+In most cases, one does not need to concern themselves with backends;
+MAGICL functionality should "just work" and dispatch to the
+appropriate backend. However, the programmer always has control, even
+dynamically in the program, of which backends should be used at a
+given time with the `magicl.backends:with-backends` macro. For instance,
+
+```
+(magicl.backends:with-backends (:blas :lisp)
+  ;; ... code ...
+  )
+```
+
+says that the code should be executed, always preferring
+`:blas`-accelerated functions, and using `:lisp`-implemented functions
+as a fall-back.
+
+```
+(magicl.backends:with-backends (:lisp)
+  ;; ... code ...
+  )
+```
+
+says to *only* use `:lisp`-implemented functions, even if other
+backends are loaded.
+
+
+The active backends can be found with the function
+`magicl.backends:active-backends`, which lists the backends to use in
+priority order.
+
+One can be even finer-grained than `with-backends`. Given a function
+`f` which has many backend implementations, one can get a specific
+implementation by using the function:
+
+```
+(magicl.backends:backend-implementation 'f :backend-name)
+```
+
+For instance
+
+```
+(magicl.backends:backend-implementation 'magicl:csd :lapack)
+```
+
+will give the implementation of the cosine-sine decomposition function
+in LAPACK. This can be called in exactly the same way `magicl:csd` can
+be called.
+
+In `backend-implementation`, if both the function name and the backend
+name are (quoted) constants, this will be looked up at compile-time,
+which is useful for writing efficient code that does not dispatch. But
+note that by doing this, `with-backends` will not be respected.
+
 ## Testing MAGICL
 
 You can run the MAGICL tests from your Lisp REPL with:
 
 ```
-(ql:quickload :magicl-tests)
 (asdf:test-system :magicl)
 ```
 
+You currently need all of the extensions working for the tests to run.
```

Comment:
Is it worth it to throw in a few easy tests that exercise `magicl/core` explicitly? More broadly, we should probably make an issue to test various backends explicitly.

Reply:
I thought in a separate test we could write test functions which iterate over the backends using `with-backends` or binding `*backends*` directly. I don't think I want to do it in this PR though. Made issue #128